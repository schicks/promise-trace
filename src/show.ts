import { Traceable } from './domain'
import { ranks, map, namer } from './utilities'

const byDependencies = (t: Traceable) => t.dependencies

const traceableRanks = ranks(byDependencies)


const nameDependencies = (t: Traceable): Map<Traceable, string> => {
    const addName = namer<Traceable>(t => t.name ?? 'missing')
    return map(
        byDependencies,
        (d): [Traceable, string] => [
            d,
            addName(d)
        ]
    )(t).reduce((acc, [d, name]) => {
        acc.set(d, name)
        return acc
    }, new Map<Traceable, string>())
}

type JsonTree = { [key in string]?: JsonTree | string | null }

const json = (
    t: Traceable,
    verbose: boolean = false
): JsonTree => {
    const names = nameDependencies(t)
    const visited = new Set<Traceable>()
    const helper = (t: Traceable): JsonTree => {
        if (byDependencies(t).length === 0) return { [names.get(t) as string]: null }
        return byDependencies(t).reduce<JsonTree>(
            (acc, d) => {
                if (!visited.has(d) || verbose) {
                    visited.add(d)
                    return { ...acc, [names.get(d) as string]: helper(d) }
                } else return {...acc, [names.get(d) as string]: null}
            },
            {}
        )
    }
    return { [names.get(t) as string]: helper(t) }
}

const list = (t: Traceable): [string, number][] => {
    const names = nameDependencies(t)
    return Array.from(traceableRanks(t).entries())
        .sort(([_, a], [__, b]) => a - b)
        .map<[string, number]>(([d, rank]) => [names.get(d) as string, rank])
}

export {
    ranks,
    list,
    json
}