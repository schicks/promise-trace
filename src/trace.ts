import { Dependencies, Traceable } from './domain'

const isTraceable = <T>(t: PromiseLike<T>): t is Traceable<T> => 'dependencies' in t
const isThenable = (t: unknown): t is PromiseLike<unknown> => t && (t as PromiseLike<unknown>).then && true

const trace = <T>(
    traceTarget: PromiseLike<T> | T,
    name?: string,
    dependencies?: Dependencies
): Traceable<T> => {
    const promise = isThenable(traceTarget)
        ? traceTarget
        : Promise.resolve(traceTarget)
    if (isTraceable(promise) && promise.name === (name ?? promise.name)) return promise
    const traced = {
        name,
        dependencies: dependencies ?? [],
        then: <U, R>(
            onFulfilled?: ((_: T) => U | PromiseLike<U>) | null,
            onRejected?: ((_: T) => R | PromiseLike<R>) | null,
            name?: string
        ): Traceable<U | R> => trace(promise.then(onFulfilled, onRejected), name, [traced])
    }
    return traced
}

export default trace