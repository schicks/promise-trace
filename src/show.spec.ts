import traceAll from './all'
import trace from './trace'
import { list, json, dot } from './show'

const line = new Array(5).fill(null).reduce(
    (acc, _, i) => acc.then(null, null, `line_${i}`),
    trace(Promise.resolve(null), 'root')
)
const T = traceAll<null>(
    Array(5).fill(null).map((_, i) => line.then(null, null, `bubble_${i}`)),
    'bubble'
)

const dags = {
    line,
    T
}


describe('visualizations', () => {
    test.each(Object.entries(dags))('it should list all the nodes in %s with their depth', (name, dag) => {
        expect(list(dag)).toMatchSnapshot()
    })

    test.each(Object.entries(dags))('it should give a json dependency tree of %s', (name, dag) => {
        expect(json(dag, true)).toMatchSnapshot()
    })

    test.each(Object.entries(dags))('it should not show the nodes of %s more than once by default', (name, dag) => {
        expect(json(dag)).toMatchSnapshot()
    })

    test.each(Object.entries(dags))('it should generate a graphviz definition for %s', (name, dag) => {
        expect(dot(dag)).toMatchSnapshot()
    })
})