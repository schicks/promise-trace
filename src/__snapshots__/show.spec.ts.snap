// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`visualizations it should give a json dependency tree of T 1`] = `
Object {
  "bubble": Object {
    "bubble_0": Object {
      "line_4": Object {
        "line_3": Object {
          "line_2": Object {
            "line_1": Object {
              "line_0": Object {
                "root": Object {
                  "root": null,
                },
              },
            },
          },
        },
      },
    },
    "bubble_1": Object {
      "line_4": Object {
        "line_3": Object {
          "line_2": Object {
            "line_1": Object {
              "line_0": Object {
                "root": Object {
                  "root": null,
                },
              },
            },
          },
        },
      },
    },
    "bubble_2": Object {
      "line_4": Object {
        "line_3": Object {
          "line_2": Object {
            "line_1": Object {
              "line_0": Object {
                "root": Object {
                  "root": null,
                },
              },
            },
          },
        },
      },
    },
    "bubble_3": Object {
      "line_4": Object {
        "line_3": Object {
          "line_2": Object {
            "line_1": Object {
              "line_0": Object {
                "root": Object {
                  "root": null,
                },
              },
            },
          },
        },
      },
    },
    "bubble_4": Object {
      "line_4": Object {
        "line_3": Object {
          "line_2": Object {
            "line_1": Object {
              "line_0": Object {
                "root": Object {
                  "root": null,
                },
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`visualizations it should give a json dependency tree of line 1`] = `
Object {
  "line_4": Object {
    "line_3": Object {
      "line_2": Object {
        "line_1": Object {
          "line_0": Object {
            "root": Object {
              "root": null,
            },
          },
        },
      },
    },
  },
}
`;

exports[`visualizations it should list all the nodes in T with their depth 1`] = `
Array [
  Array [
    "bubble",
    0,
  ],
  Array [
    "bubble_0",
    1,
  ],
  Array [
    "bubble_1",
    1,
  ],
  Array [
    "bubble_2",
    1,
  ],
  Array [
    "bubble_3",
    1,
  ],
  Array [
    "bubble_4",
    1,
  ],
  Array [
    "line_4",
    2,
  ],
  Array [
    "line_3",
    3,
  ],
  Array [
    "line_2",
    4,
  ],
  Array [
    "line_1",
    5,
  ],
  Array [
    "line_0",
    6,
  ],
  Array [
    "root",
    7,
  ],
]
`;

exports[`visualizations it should list all the nodes in line with their depth 1`] = `
Array [
  Array [
    "line_4",
    0,
  ],
  Array [
    "line_3",
    1,
  ],
  Array [
    "line_2",
    2,
  ],
  Array [
    "line_1",
    3,
  ],
  Array [
    "line_0",
    4,
  ],
  Array [
    "root",
    5,
  ],
]
`;

exports[`visualizations it should not show the nodes of T more than once by default 1`] = `
Object {
  "bubble": Object {
    "bubble_0": Object {
      "line_4": Object {
        "line_3": Object {
          "line_2": Object {
            "line_1": Object {
              "line_0": Object {
                "root": Object {
                  "root": null,
                },
              },
            },
          },
        },
      },
    },
    "bubble_1": Object {
      "line_4": null,
    },
    "bubble_2": Object {
      "line_4": null,
    },
    "bubble_3": Object {
      "line_4": null,
    },
    "bubble_4": Object {
      "line_4": null,
    },
  },
}
`;

exports[`visualizations it should not show the nodes of line more than once by default 1`] = `
Object {
  "line_4": Object {
    "line_3": Object {
      "line_2": Object {
        "line_1": Object {
          "line_0": Object {
            "root": Object {
              "root": null,
            },
          },
        },
      },
    },
  },
}
`;
